src/azul/portal_service.py

```diff
@@ -0,0 +1,167 @@
+from copy import deepcopy
+from typing import (
+    Callable,
+    Optional,
+)
+import json
+import logging
+
+import boto3
+
+from azul import config
+from azul.plugin import Plugin
+from azul.types import JSONs
+from azul.version_service import (
+    VersionService,
+    VersionConflict,
+    NoSuchObjectVersion,
+)
+
+log = logging.getLogger(__name__)
+
+
+class PortalService:
+
+    def __init__(self):
+        self.client = boto3.client('s3')
+        self.version_service = VersionService()
+
+    def crud(self, operation: Callable[[JSONs], Optional[JSONs]]) -> None:
+        """
+        Perform a concurrent read/write operation on the portal integrations DB.
+
+        :param operation: Callable that accepts the latest version of the portal
+        DB and optionally returns an updated version.
+        """
+        while True:
+            version = self.version_service.get(self._db_url())
+            if version is None:
+                log.info('Portal integration DB not found in S3; uploading hard-coded DB.')
+                self._create_db()
+                continue
+            else:
+                try:
+                    db = self._read_db(version)
+                except NoSuchObjectVersion:
+                    # Wait for latest version to appear in S3.
+                    continue
+            db = operation(db)
+            if db is None:
+                # Operation is read-only; we're done.
+                break
+            else:
+                new_version = self._write_db(db)
+                if self._update_version(version, new_version):
+                    break
+                else:
+                    # Retry with up-to-date DB
+                    continue
+
+    def _update_version(self, version: Optional[str], new_version: str) -> bool:
+        """
+        Try to update the most current version of the S3 DB and report success.
+        Update is rejected if `version` is not current.
+        :param version: the current version.
+        :param new_version: the replacement version.
+        :return: whether hte current version was updated.
+        """
+        try:
+            self.version_service.put(self._db_url(), version, new_version)
+        except VersionConflict:
+            # Operation was performed on outdated DB and now erroneously
+            # exists in S3.
+            self._delete_db(new_version)
+            return False
+        else:
+            return True
+
+    def _create_db(self) -> None:
+        """
+        Write hardcoded portal integrations DB to S3 if no other DB is present.
+        """
+        plugin = Plugin.load()
+        db = self.demultiplex(plugin.portal_integrations_db())
+        init_version = self._write_db(db)
+        self._update_version(None, init_version)
+        # Ignore failure, just means that someone else created it first
```

Comment:
This race should not be ignored and a retry should be triggered.

Reply:
https://github.com/DataBiosphere/azul/pull/1428#discussion_r348674257

Let's keep this in one thread